File: cpp/ql/src/Security/CWE/CWE-020/ExternalAPITaintStepExample.cpp
#include <cstdio>

void do_get(FILE* request, FILE* response) {
  char user_id[1024];
  fgets(user_id, 1024, request);

  char buffer[1024];
  strcat(buffer, "SELECT * FROM user WHERE user_id='");
  strcat(buffer, user_id);
  strcat(buffer, "'");

  // ...
}

File: java/ql/src/Security/CWE/CWE-020/ExternalAPITaintStepExample.java
public class SQLInjection extends HttpServlet {
	protected void doGet(HttpServletRequest request, HttpServletResponse response)
	throws ServletException, IOException {

		StringBuilder sqlQueryBuilder = new StringBuilder();
		sqlQueryBuilder.append("SELECT * FROM user WHERE user_id='");
		// BAD: a request parameter is concatenated directly into a SQL query
		sqlQueryBuilder.append(request.getParameter("user_id"));
		sqlQueryBuilder.append("'");

		// ...
	}
}
